File: src/tinymonitor/alerts/gotify.py
import requests
from datetime import datetime
from tinymonitor.alerts.provider import AlertProvider, logger
from tinymonitor.utils import system_info

class GotifyAlert(AlertProvider):
    def send(self, component, level, value, title, message):
        server_url = self.config.get("url")
        token = self.config.get("token")
        
        if not server_url or not token:
            logger.error(f"[{self.name}] No url or token provided")
            return

        # Ensure URL ends with /message
        if not server_url.endswith("/message"):
            if not server_url.endswith("/"):
                server_url += "/"
            server_url += "message"

        # Gotify priority mapping (0-10)
        if level == "CRITICAL":
            priority = 8
        elif level == "WARNING":
            priority = 5
        else:
            priority = 2

        # System Info
        hostname = system_info.get_hostname()
        execution_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        ip_private = system_info.get_private_ip()
        ip_public = system_info.get_public_ip()
        load_avg = system_info.get_load_avg()
        uptime_pretty = system_info.get_uptime()
        
        # Enriched message construction (Markdown supported by Gotify)
        # Note: We use two spaces at the end of lines to force a line break in Markdown
        full_message = (
            f"**Component** : {component}  \n"
            f"**Value**     : {value}  \n"
            f"**Level**     : {level}  \n\n"
            f"__Machine Context__  \n"
            f"🖥️ **Server**    : `{hostname}`  \n"
            f"🏠 **Private IP**: `{ip_private}`  \n"
            f"🌍 **Public IP** : `{ip_public}`  \n"
            f"⚙️ **Load Avg**  : `{load_avg}`  \n"
            f"⏱️ **Uptime**    : `{uptime_pretty}`  \n"
            f"🕒 **Time**      : {execution_time}"
        )

        payload = {
            "title": title,
            "message": full_message,
            "priority": priority,
            "extras": {
                "client::display": {
                    "contentType": "text/markdown"
                }
            }
        }

        headers = {
            "X-Gotify-Key": token
        }

        try:
            response = requests.post(server_url, json=payload, headers=headers, timeout=10)
            if response.ok:
                logger.info(f"[{self.name}] Alert sent successfully")
            else:
                logger.error(f"[{self.name}] Failed to send alert: {response.status_code} {response.text}")
        except Exception as e:
            logger.error(f"[{self.name}] Error sending alert: {e}")

File: internal/alerts/gotify.go
package alerts

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/Gu1llaum-3/tinymonitor/internal/config"
	"github.com/Gu1llaum-3/tinymonitor/internal/models"
	"github.com/Gu1llaum-3/tinymonitor/internal/utils"
)

// GotifyProvider sends alerts to Gotify
type GotifyProvider struct {
	BaseProvider
	url   string
	token string
}

// NewGotifyProvider creates a new Gotify provider
func NewGotifyProvider(cfg config.GotifyConfig) *GotifyProvider {
	return &GotifyProvider{
		BaseProvider: BaseProvider{
			ProviderName: "gotify",
			Enabled:      cfg.Enabled,
			Levels:       cfg.Levels,
			Rules:        cfg.Rules,
		},
		url:   cfg.URL,
		token: cfg.Token,
	}
}

// Send sends an alert to Gotify
func (p *GotifyProvider) Send(alert models.Alert) error {
	if p.url == "" || p.token == "" {
		return fmt.Errorf("no url or token provided")
	}

	// Ensure URL ends with /message
	serverURL := p.url
	if !strings.HasSuffix(serverURL, "/message") {
		if !strings.HasSuffix(serverURL, "/") {
			serverURL += "/"
		}
		serverURL += "message"
	}

	// Priority mapping
	var priority int
	switch alert.Level {
	case models.SeverityCritical:
		priority = 8
	case models.SeverityWarning:
		priority = 5
	case models.SeverityRecovery:
		priority = 3
	default:
		priority = 2
	}

	// System Info
	hostname := utils.GetHostname()
	executionTime := time.Now().Format("2006-01-02 15:04:05")
	ipPrivate := utils.GetPrivateIP()
	ipPublic := utils.GetPublicIP()
	loadAvg := utils.GetLoadAvg()
	uptimePretty := utils.GetUptime()

	// Enriched message (Markdown supported)
	fullMessage := fmt.Sprintf(`**Component** : %s
**Value**     : %s
**Level**     : %s

__Machine Context__
🖥️ **Server**    : `+"`%s`"+`
🏠 **Private IP**: `+"`%s`"+`
🌍 **Public IP** : `+"`%s`"+`
⚙️ **Load Avg**  : `+"`%s`"+`
⏱️ **Uptime**    : `+"`%s`"+`
🕒 **Time**      : %s`,
		alert.Component, alert.Value, alert.Level,
		hostname, ipPrivate, ipPublic, loadAvg, uptimePretty, executionTime)

	payload := map[string]interface{}{
		"title":    alert.Title,
		"message":  fullMessage,
		"priority": priority,
		"extras": map[string]interface{}{
			"client::display": map[string]string{
				"contentType": "text/markdown",
			},
		},
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("POST", serverURL, bytes.NewReader(body))
	if err != nil {
		return err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Gotify-Key", p.token)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("failed to send alert: status %d", resp.StatusCode)
	}

	LogInfo(p.ProviderName, "Alert sent successfully")
	return nil
}
